// twelvelabs-js/dist/wrapper/resources/IndexesVideosWrapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexesVideosWrapper = void 0;
const Client_1 = require("../../api/resources/indexes/resources/videos/client/Client");
const TwelvelabsApi = __importStar(require("../../api/index"));
const core = __importStar(require("../../core"));
const environments = __importStar(require("../../environments"));
const url_join_1 = __importDefault(require("url-join"));
const serializers = __importStar(require("../../serialization/index"));
const errors = __importStar(require("../../errors/index"));
const version_1 = require("../../version");
/**
 * Wrapper for the IndexesVideos client that adds additional functionality.
 */
class IndexesVideosWrapper extends Client_1.Videos {
    constructor(options) {
        super(options);
    }
    /**
     * This method returns a list of the videos in the specified index. By default, the API returns your videos sorted by creation date, with the newest at the top of the list.
     *
     * @param {string} indexId - The unique identifier of the index for which the API will retrieve the videos.
     * @param {TwelvelabsApi.indexes.VideosListRequest} request
     * @param {Videos.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     *
     * @example
     *     await client.indexes.videos.list("6298d673f1090f1100476d4c", {
     *         page: 1,
     *         pageLimit: 10,
     *         sortBy: "created_at",
     *         sortOption: "desc",
     *         filename: "01.mp4",
     *         createdAt: "2024-08-16T16:53:59Z",
     *         updatedAt: "2024-08-16T16:53:59Z"
     *     })
     */
    list(indexId_1) {
        return __awaiter(this, arguments, void 0, function* (indexId, request = {}, requestOptions) {
            const list = core.HttpResponsePromise.interceptFunction((request) => __awaiter(this, void 0, void 0, function* () {
                var _a, _b;
                const { page, pageLimit, sortBy, sortOption, filename, duration, fps, width, height, size, createdAt, updatedAt, userMetadata, } = request;
                const _queryParams = {};
                if (page != null) {
                    _queryParams["page"] = page.toString();
                }
                if (pageLimit != null) {
                    _queryParams["page_limit"] = pageLimit.toString();
                }
                if (sortBy != null) {
                    _queryParams["sort_by"] = sortBy;
                }
                if (sortOption != null) {
                    _queryParams["sort_option"] = sortOption;
                }
                if (filename != null) {
                    _queryParams["filename"] = filename;
                }
                if (duration != null) {
                    _queryParams["duration"] = duration.toString();
                }
                if (fps != null) {
                    _queryParams["fps"] = fps.toString();
                }
                if (width != null) {
                    _queryParams["width"] = width.toString();
                }
                if (height != null) {
                    _queryParams["height"] = height.toString();
                }
                if (size != null) {
                    _queryParams["size"] = size.toString();
                }
                if (createdAt != null) {
                    _queryParams["created_at"] = createdAt;
                }
                if (updatedAt != null) {
                    _queryParams["updated_at"] = updatedAt;
                }
                // Flatten userMetadata instead of using toJson
                if (userMetadata != null && typeof userMetadata === "object") {
                    for (const [key, value] of Object.entries(userMetadata)) {
                        if (value != null) {
                            _queryParams[key] = typeof value === "string" ? value : String(value);
                        }
                    }
                }
                const _response = yield core.fetcher({
                    url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, `indexes/${encodeURIComponent(indexId)}/videos`),
                    method: "GET",
                    headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": version_1.SDK_VERSION, "User-Agent": "twelvelabs-js/" + version_1.SDK_VERSION, "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                    contentType: "application/json",
                    queryParameters: _queryParams,
                    requestType: "json",
                    timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                    maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                    abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                });
                if (_response.ok) {
                    return {
                        data: serializers.indexes.VideosListResponse.parseOrThrow(_response.body, {
                            unrecognizedObjectKeys: "passthrough",
                            allowUnrecognizedUnionMembers: true,
                            allowUnrecognizedEnumValues: true,
                            breadcrumbsPrefix: ["response"],
                        }),
                        rawResponse: _response.rawResponse,
                    };
                }
                if (_response.error.reason === "status-code") {
                    switch (_response.error.statusCode) {
                        case 400:
                            throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                        default:
                            throw new errors.TwelvelabsApiError({
                                statusCode: _response.error.statusCode,
                                body: _response.error.body,
                                rawResponse: _response.rawResponse,
                            });
                    }
                }
                switch (_response.error.reason) {
                    case "non-json":
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.rawBody,
                            rawResponse: _response.rawResponse,
                        });
                    case "timeout":
                        throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling GET /indexes/{index-id}/videos.");
                    case "unknown":
                        throw new errors.TwelvelabsApiError({
                            message: _response.error.errorMessage,
                            rawResponse: _response.rawResponse,
                        });
                }
            }));
            let _offset = (request === null || request === void 0 ? void 0 : request.page) != null ? request === null || request === void 0 ? void 0 : request.page : 1;
            const dataWithRawResponse = yield list(request).withRawResponse();
            return new core.Pageable({
                response: dataWithRawResponse.data,
                rawResponse: dataWithRawResponse.rawResponse,
                hasNextPage: (response) => { var _a; return ((_a = response === null || response === void 0 ? void 0 : response.data) !== null && _a !== void 0 ? _a : []).length > 0; },
                getItems: (response) => { var _a; return (_a = response === null || response === void 0 ? void 0 : response.data) !== null && _a !== void 0 ? _a : []; },
                loadPage: (_response) => {
                    _offset += 1;
                    return list(core.setObjectProperty(request, "page", _offset));
                },
            });
        });
    }
}
exports.IndexesVideosWrapper = IndexesVideosWrapper;
