// twelvelabs-js/dist/wrapper/resources/SearchWrapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchWrapper = void 0;
const Client_1 = require("../../api/resources/search/client/Client");
const core = __importStar(require("../../core"));
class SearchWrapper extends Client_1.Search {
    constructor(options) {
        super(options);
    }
    /**
     * Use this endpoint to search for relevant matches in an index using text, media, or a combination of both as your query.
     *
     * **Text queries**:
     * - Use the `query_text` parameter to specify your query.
     *
     * **Media queries**:
     * - Set the `query_media_type` parameter to the corresponding media type (example: `image`).
     * - Specify either one of the following parameters:
     *   - `query_media_url`: Publicly accessible URL of your media file.
     *   - `query_media_file`: Local media file.
     *   If both `query_media_url` and `query_media_file` are specified in the same request, `query_media_url` takes precedence.
     *
     * **Composed text and media queries** (Marengo 3.0 only):
     * - Use the `query_text` parameter for your text query.
     * - Set `query_media_type` to `image`.
     * - Specify the image using either the `query_media_url` or the `query_media_file` parameter.
     *
     *   Example: Provide an image of a car and include  "red color"  in your query to find red instances of that car model.
     *
     * <Note title="Note">
     *   When using images in your search queries (either as media queries or in composed searches), ensure your image files meet the [format requirements](/v1.3/docs/concepts/models/marengo#image-file-requirements).
     * </Note>
     *
     * <Note title="Note">
     * This endpoint is rate-limited. For details, see the [Rate limits](/v1.3/docs/get-started/rate-limits) page.
     * </Note>
     *
     * @param {TwelvelabsApi.SearchCreateRequest} request
     * @param {Search.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     * @throws {@link TwelvelabsApi.TooManyRequestsError}
     *
     * @example
     *     await client.search.create({
     *         indexId: "index_id",
     *         searchOptions: ["visual"]
     *     })
     */
    query(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.create(request, requestOptions);
            const retrieveRequest = {};
            if (request.includeUserMetadata) {
                retrieveRequest.includeUserMetadata = request.includeUserMetadata;
            }
            return new core.Pageable({
                response,
                rawResponse: {},
                hasNextPage: (response) => { var _a; return ((_a = response === null || response === void 0 ? void 0 : response.pageInfo) === null || _a === void 0 ? void 0 : _a.nextPageToken) != null; },
                getItems: (response) => { var _a; return (_a = response === null || response === void 0 ? void 0 : response.data) !== null && _a !== void 0 ? _a : []; },
                loadPage: (response) => {
                    return this.retrieve(response.pageInfo.nextPageToken, retrieveRequest, requestOptions);
                },
            });
        });
    }
}
exports.SearchWrapper = SearchWrapper;
