// twelvelabs-js/dist/serialization/types/AudioInputRequest.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioInputRequest = void 0;
const core = __importStar(require("../../core"));
const MediaSource_1 = require("./MediaSource");
const AudioSegmentation_1 = require("./AudioSegmentation");
const AudioInputRequestEmbeddingOptionItem_1 = require("./AudioInputRequestEmbeddingOptionItem");
const AudioInputRequestEmbeddingScopeItem_1 = require("./AudioInputRequestEmbeddingScopeItem");
exports.AudioInputRequest = core.serialization.object({
    mediaSource: core.serialization.property("media_source", MediaSource_1.MediaSource),
    startSec: core.serialization.property("start_sec", core.serialization.number().optional()),
    endSec: core.serialization.property("end_sec", core.serialization.number().optional()),
    segmentation: AudioSegmentation_1.AudioSegmentation.optional(),
    embeddingOption: core.serialization.property("embedding_option", core.serialization.list(AudioInputRequestEmbeddingOptionItem_1.AudioInputRequestEmbeddingOptionItem).optional()),
    embeddingScope: core.serialization.property("embedding_scope", core.serialization.list(AudioInputRequestEmbeddingScopeItem_1.AudioInputRequestEmbeddingScopeItem).optional()),
});
