// twelvelabs-js/dist/serialization/types/GetUploadStatusResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetUploadStatusResponse = void 0;
const core = __importStar(require("../../core"));
const MultipartUploadStatusType_1 = require("./MultipartUploadStatusType");
const ChunkInfo_1 = require("./ChunkInfo");
const PageInfo_1 = require("./PageInfo");
const CreatedAt_1 = require("./CreatedAt");
const UpdatedAt_1 = require("./UpdatedAt");
exports.GetUploadStatusResponse = core.serialization.object({
    uploadId: core.serialization.property("upload_id", core.serialization.string().optional()),
    status: MultipartUploadStatusType_1.MultipartUploadStatusType.optional(),
    uploadedChunks: core.serialization.property("uploaded_chunks", core.serialization.list(ChunkInfo_1.ChunkInfo).optional()),
    chunksCompleted: core.serialization.property("chunks_completed", core.serialization.number().optional()),
    chunksFailed: core.serialization.property("chunks_failed", core.serialization.number().optional()),
    chunksPending: core.serialization.property("chunks_pending", core.serialization.number().optional()),
    pageInfo: core.serialization.property("page_info", PageInfo_1.PageInfo.optional()),
    totalSize: core.serialization.property("total_size", core.serialization.number().optional()),
    uploadedSize: core.serialization.property("uploaded_size", core.serialization.number().optional()),
    createdAt: core.serialization.property("created_at", CreatedAt_1.CreatedAt.optional()),
    updatedAt: core.serialization.property("updated_at", UpdatedAt_1.UpdatedAt.optional()),
});
