// twelvelabs-js/dist/api/resources/indexes/resources/indexedAssets/client/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexedAssets = void 0;
const environments = __importStar(require("../../../../../../environments"));
const core = __importStar(require("../../../../../../core"));
const TwelvelabsApi = __importStar(require("../../../../../index"));
const serializers = __importStar(require("../../../../../../serialization/index"));
const json_1 = require("../../../../../../core/json");
const url_join_1 = __importDefault(require("url-join"));
const errors = __importStar(require("../../../../../../errors/index"));
class IndexedAssets {
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * This method returns a list of the indexed assets in the specified index. By default, the platform returns your indexed assets sorted by creation date, with the newest at the top of the list.
     *
     * @param {string} indexId - The unique identifier of the index for which the platform will retrieve the indexed assets.
     * @param {TwelvelabsApi.indexes.IndexedAssetsListRequest} request
     * @param {IndexedAssets.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     *
     * @example
     *     await client.indexes.indexedAssets.list("6298d673f1090f1100476d4c", {
     *         page: 1,
     *         pageLimit: 10,
     *         sortBy: "created_at",
     *         sortOption: "desc",
     *         filename: "01.mp4",
     *         duration: 1.1,
     *         fps: 1.1,
     *         width: 1.1,
     *         height: 1,
     *         size: 1.1,
     *         createdAt: "2024-08-16T16:53:59Z",
     *         updatedAt: "2024-08-16T16:53:59Z"
     *     })
     */
    list(indexId_1) {
        return __awaiter(this, arguments, void 0, function* (indexId, request = {}, requestOptions) {
            const list = core.HttpResponsePromise.interceptFunction((request) => __awaiter(this, void 0, void 0, function* () {
                var _a, _b;
                const { page, pageLimit, sortBy, sortOption, status, filename, duration, fps, width, height, size, createdAt, updatedAt, userMetadata, } = request;
                const _queryParams = {};
                if (page != null) {
                    _queryParams["page"] = page.toString();
                }
                if (pageLimit != null) {
                    _queryParams["page_limit"] = pageLimit.toString();
                }
                if (sortBy != null) {
                    _queryParams["sort_by"] = sortBy;
                }
                if (sortOption != null) {
                    _queryParams["sort_option"] = sortOption;
                }
                if (status != null) {
                    if (Array.isArray(status)) {
                        _queryParams["status"] = status.map((item) => serializers.indexes.IndexedAssetsListRequestStatusItem.jsonOrThrow(item, {
                            unrecognizedObjectKeys: "strip",
                        }));
                    }
                    else {
                        _queryParams["status"] = serializers.indexes.IndexedAssetsListRequestStatusItem.jsonOrThrow(status, { unrecognizedObjectKeys: "strip" });
                    }
                }
                if (filename != null) {
                    _queryParams["filename"] = filename;
                }
                if (duration != null) {
                    _queryParams["duration"] = duration.toString();
                }
                if (fps != null) {
                    _queryParams["fps"] = fps.toString();
                }
                if (width != null) {
                    _queryParams["width"] = width.toString();
                }
                if (height != null) {
                    _queryParams["height"] = height.toString();
                }
                if (size != null) {
                    _queryParams["size"] = size.toString();
                }
                if (createdAt != null) {
                    _queryParams["created_at"] = createdAt;
                }
                if (updatedAt != null) {
                    _queryParams["updated_at"] = updatedAt;
                }
                if (userMetadata != null) {
                    _queryParams["user_metadata"] = (0, json_1.toJson)(userMetadata);
                }
                const _response = yield core.fetcher({
                    url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, `indexes/${encodeURIComponent(indexId)}/indexed-assets`),
                    method: "GET",
                    headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": "1.1.0", "User-Agent": "twelvelabs-js/1.1.0", "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                    contentType: "application/json",
                    queryParameters: _queryParams,
                    requestType: "json",
                    timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                    maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                    abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                });
                if (_response.ok) {
                    return {
                        data: serializers.indexes.IndexedAssetsListResponse.parseOrThrow(_response.body, {
                            unrecognizedObjectKeys: "passthrough",
                            allowUnrecognizedUnionMembers: true,
                            allowUnrecognizedEnumValues: true,
                            breadcrumbsPrefix: ["response"],
                        }),
                        rawResponse: _response.rawResponse,
                    };
                }
                if (_response.error.reason === "status-code") {
                    switch (_response.error.statusCode) {
                        case 400:
                            throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                        default:
                            throw new errors.TwelvelabsApiError({
                                statusCode: _response.error.statusCode,
                                body: _response.error.body,
                                rawResponse: _response.rawResponse,
                            });
                    }
                }
                switch (_response.error.reason) {
                    case "non-json":
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.rawBody,
                            rawResponse: _response.rawResponse,
                        });
                    case "timeout":
                        throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling GET /indexes/{index-id}/indexed-assets.");
                    case "unknown":
                        throw new errors.TwelvelabsApiError({
                            message: _response.error.errorMessage,
                            rawResponse: _response.rawResponse,
                        });
                }
            }));
            let _offset = (request === null || request === void 0 ? void 0 : request.page) != null ? request === null || request === void 0 ? void 0 : request.page : 1;
            const dataWithRawResponse = yield list(request).withRawResponse();
            return new core.Pageable({
                response: dataWithRawResponse.data,
                rawResponse: dataWithRawResponse.rawResponse,
                hasNextPage: (response) => { var _a; return ((_a = response === null || response === void 0 ? void 0 : response.data) !== null && _a !== void 0 ? _a : []).length > 0; },
                getItems: (response) => { var _a; return (_a = response === null || response === void 0 ? void 0 : response.data) !== null && _a !== void 0 ? _a : []; },
                loadPage: (_response) => {
                    _offset += 1;
                    return list(core.setObjectProperty(request, "page", _offset));
                },
            });
        });
    }
    /**
     * This method indexes an uploaded asset to make it searchable and analyzable. Indexing processes your content and extracts information that enables the platform to search and analyze your videos.
     *
     * This operation is asynchronous. The platform returns an indexed asset ID immediately and processes your content in the background. Monitor the indexing status to know when your content is ready to use.
     *
     * Your asset must meet the requirements based on your workflow:
     * - **Search**: [Marengo requirements](/v1.3/docs/concepts/models/marengo#video-file-requirements)
     * - **Video analysis**: [Pegasus requirements](/v1.3/docs/concepts/models/pegasus#input-requirements).
     *
     * If you want to both search and analyze your videos, the most restrictive requirements apply.
     *
     * @param {string} indexId - The unique identifier of the index to which the asset will be indexed.
     * @param {TwelvelabsApi.indexes.IndexedAssetsCreateRequest} request
     * @param {IndexedAssets.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     * @throws {@link TwelvelabsApi.NotFoundError}
     * @throws {@link TwelvelabsApi.InternalServerError}
     *
     * @example
     *     await client.indexes.indexedAssets.create("6298d673f1090f1100476d4c", {
     *         assetId: "6298d673f1090f1100476d4c"
     *     })
     */
    create(indexId, request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__create(indexId, request, requestOptions));
    }
    __create(indexId, request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const _response = yield core.fetcher({
                url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, `indexes/${encodeURIComponent(indexId)}/indexed-assets`),
                method: "POST",
                headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": "1.1.0", "User-Agent": "twelvelabs-js/1.1.0", "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                contentType: "application/json",
                requestType: "json",
                body: serializers.indexes.IndexedAssetsCreateRequest.jsonOrThrow(request, {
                    unrecognizedObjectKeys: "strip",
                }),
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return {
                    data: serializers.indexes.IndexedAssetsCreateResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new TwelvelabsApi.NotFoundError(_response.error.body, _response.rawResponse);
                    case 500:
                        throw new TwelvelabsApi.InternalServerError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TwelvelabsApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling POST /indexes/{index-id}/indexed-assets.");
                case "unknown":
                    throw new errors.TwelvelabsApiError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    /**
     * This method retrieves information about an indexed asset, including its status, metadata, and optional embeddings or transcription.
     *
     * **Common use cases**:
     *
     * - Monitor indexing progress:
     *   - Call this endpoint after creating an indexed asset
     *   - Check the `status` field until it shows `ready`
     *   - Once ready, your content is available for search and analysis
     *
     * - Retrieve  asset metadata:
     *   - Retrieve system metadata (duration, resolution, filename)
     *   - Access user-defined metadata
     *
     * - Retrieve embeddings:
     *   - Include the `embedding_option` parameter to retrieve video embeddings
     *   - Requires the Marengo video understanding model to be enabled in your index
     *
     * - Retrieve transcriptions:
     *   - Set the `transcription` parameter to `true` to retrieve spoken words from your video
     *
     * @param {string} indexId - The unique identifier of the index to which the indexed asset has been uploaded.
     * @param {string} indexedAssetId - The unique identifier of the indexed asset to retrieve.
     * @param {TwelvelabsApi.indexes.IndexedAssetsRetrieveRequest} request
     * @param {IndexedAssets.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     * @throws {@link TwelvelabsApi.NotFoundError}
     *
     * @example
     *     await client.indexes.indexedAssets.retrieve("6298d673f1090f1100476d4c", "6298d673f1090f1100476d4c", {
     *         transcription: true
     *     })
     */
    retrieve(indexId, indexedAssetId, request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__retrieve(indexId, indexedAssetId, request, requestOptions));
    }
    __retrieve(indexId_1, indexedAssetId_1) {
        return __awaiter(this, arguments, void 0, function* (indexId, indexedAssetId, request = {}, requestOptions) {
            var _a, _b;
            const { embeddingOption, transcription } = request;
            const _queryParams = {};
            if (embeddingOption != null) {
                if (Array.isArray(embeddingOption)) {
                    _queryParams["embedding_option"] = embeddingOption.map((item) => serializers.indexes.IndexedAssetsRetrieveRequestEmbeddingOptionItem.jsonOrThrow(item, {
                        unrecognizedObjectKeys: "strip",
                    }));
                }
                else {
                    _queryParams["embedding_option"] =
                        serializers.indexes.IndexedAssetsRetrieveRequestEmbeddingOptionItem.jsonOrThrow(embeddingOption, {
                            unrecognizedObjectKeys: "strip",
                        });
                }
            }
            if (transcription != null) {
                _queryParams["transcription"] = transcription.toString();
            }
            const _response = yield core.fetcher({
                url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, `indexes/${encodeURIComponent(indexId)}/indexed-assets/${encodeURIComponent(indexedAssetId)}`),
                method: "GET",
                headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": "1.1.0", "User-Agent": "twelvelabs-js/1.1.0", "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                contentType: "application/json",
                queryParameters: _queryParams,
                requestType: "json",
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return {
                    data: serializers.IndexedAssetDetailed.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new TwelvelabsApi.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TwelvelabsApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling GET /indexes/{index-id}/indexed-assets/{indexed-asset-id}.");
                case "unknown":
                    throw new errors.TwelvelabsApiError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    /**
     * This method deletes all the information about the specified indexed asset. This action cannot be undone.
     *
     * @param {string} indexId - The unique identifier of the index to which the indexed asset has been uploaded.
     * @param {string} indexedAssetId - The unique identifier of the indexed asset to delete.
     * @param {IndexedAssets.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     *
     * @example
     *     await client.indexes.indexedAssets.delete("6298d673f1090f1100476d4c", "6298d673f1090f1100476d4c")
     */
    delete(indexId, indexedAssetId, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__delete(indexId, indexedAssetId, requestOptions));
    }
    __delete(indexId, indexedAssetId, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const _response = yield core.fetcher({
                url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, `indexes/${encodeURIComponent(indexId)}/indexed-assets/${encodeURIComponent(indexedAssetId)}`),
                method: "DELETE",
                headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": "1.1.0", "User-Agent": "twelvelabs-js/1.1.0", "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                contentType: "application/json",
                requestType: "json",
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return { data: undefined, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TwelvelabsApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling DELETE /indexes/{index-id}/indexed-assets/{indexed-asset-id}.");
                case "unknown":
                    throw new errors.TwelvelabsApiError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    /**
     * Use this method to update one or more fields of the metadata of an indexed asset. Also, can delete a field by setting it to null.
     *
     * @param {string} indexId - The unique identifier of the index to which the indexed asset has been uploaded.
     * @param {string} indexedAssetId - The unique identifier of the indexed asset to update.
     * @param {TwelvelabsApi.indexes.IndexedAssetsUpdateRequest} request
     * @param {IndexedAssets.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     *
     * @example
     *     await client.indexes.indexedAssets.update("6298d673f1090f1100476d4c", "6298d673f1090f1100476d4c", {
     *         userMetadata: {
     *             "category": "recentlyAdded",
     *             "batchNumber": 5,
     *             "rating": 9.3,
     *             "needsReview": true
     *         }
     *     })
     */
    update(indexId, indexedAssetId, request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__update(indexId, indexedAssetId, request, requestOptions));
    }
    __update(indexId_1, indexedAssetId_1) {
        return __awaiter(this, arguments, void 0, function* (indexId, indexedAssetId, request = {}, requestOptions) {
            var _a, _b;
            const _response = yield core.fetcher({
                url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, `indexes/${encodeURIComponent(indexId)}/indexed-assets/${encodeURIComponent(indexedAssetId)}`),
                method: "PATCH",
                headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": "1.1.0", "User-Agent": "twelvelabs-js/1.1.0", "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                contentType: "application/json",
                requestType: "json",
                body: serializers.indexes.IndexedAssetsUpdateRequest.jsonOrThrow(request, {
                    unrecognizedObjectKeys: "strip",
                }),
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return { data: undefined, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TwelvelabsApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling PATCH /indexes/{index-id}/indexed-assets/{indexed-asset-id}.");
                case "unknown":
                    throw new errors.TwelvelabsApiError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    _getCustomAuthorizationHeaders() {
        return __awaiter(this, void 0, void 0, function* () {
            const apiKeyValue = yield core.Supplier.get(this._options.apiKey);
            return { "x-api-key": apiKeyValue };
        });
    }
}
exports.IndexedAssets = IndexedAssets;
