// twelvelabs-js/serialization/types/SearchItemClipsItem.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchItemClipsItem = void 0;
const core = __importStar(require("../../core"));
const ScoreSearchTerms_1 = require("./ScoreSearchTerms");
const StartTime_1 = require("./StartTime");
const EndTime_1 = require("./EndTime");
const Confidence_1 = require("./Confidence");
const Rank_1 = require("./Rank");
const ThumbnailUrl_1 = require("./ThumbnailUrl");
const UserMetadata_1 = require("./UserMetadata");
exports.SearchItemClipsItem = core.serialization.object({
    score: ScoreSearchTerms_1.ScoreSearchTerms.optional(),
    start: StartTime_1.StartTime.optional(),
    end: EndTime_1.EndTime.optional(),
    confidence: Confidence_1.Confidence.optional(),
    rank: Rank_1.Rank.optional(),
    thumbnailUrl: core.serialization.property("thumbnail_url", ThumbnailUrl_1.ThumbnailUrl.optional()),
    transcription: core.serialization.string().optional(),
    videoId: core.serialization.property("video_id", core.serialization.string().optional()),
    userMetadata: core.serialization.property("user_metadata", UserMetadata_1.UserMetadata.optional()),
});
