// twelvelabs-js/wrapper/TwelveLabs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TwelveLabs = void 0;
const Client_1 = require("../Client");
const SearchWrapper_1 = require("./resources/SearchWrapper");
const TasksWrapper_1 = require("./resources/TasksWrapper");
const EmbedWrapper_1 = require("./resources/EmbedWrapper");
const IndexesWrapper_1 = require("./resources/IndexesWrapper");
const MultipartUploadWrapper_1 = require("./resources/MultipartUploadWrapper");
const __1 = require("..");
const core = __importStar(require("../core"));
class TwelveLabs {
    constructor({ apiKey = process.env.TWELVE_LABS_API_KEY }) {
        if (!apiKey) {
            throw new Error("Provide `apiKey` to initialize a client or set the TWELVE_LABS_API_KEY environment variable. You can see the API Key in the Dashboard page: https://dashboard.playground.io");
        }
        const clientOptions = {
            apiKey: apiKey,
            environment: process.env.TWELVELABS_BASE_URL || __1.TwelvelabsApiEnvironment.Default,
        };
        this._options = clientOptions;
        this.client = new Client_1.TwelvelabsApiClient(clientOptions);
        this.summarize = this.client.summarize.bind(this.client);
        this.gist = this.client.gist.bind(this.client);
        this.analyze = this.client.analyze.bind(this.client);
        this.analyzeStream = this.client.analyzeStream.bind(this.client);
        this.assets = this.client.assets;
        this.entityCollections = this.client.entityCollections;
        // Use custom wrappers instead of default clients
        this.tasks = new TasksWrapper_1.TasksWrapper(clientOptions);
        this.indexes = new IndexesWrapper_1.IndexesWrapper(clientOptions);
        this.search = new SearchWrapper_1.SearchWrapper(clientOptions);
        this.embed = new EmbedWrapper_1.EmbedWrapper(clientOptions);
        this.multipartUpload = new MultipartUploadWrapper_1.MultipartUploadWrapper(clientOptions);
    }
    _getCustomAuthorizationHeaders() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const apiKeyValue = (_a = (yield core.Supplier.get(this._options.apiKey))) !== null && _a !== void 0 ? _a : process === null || process === void 0 ? void 0 : process.env["TWELVE_LABS_API_KEY"];
            return { "x-api-key": apiKeyValue };
        });
    }
}
exports.TwelveLabs = TwelveLabs;
