// twelvelabs-js/api/resources/embed/resources/v2/client/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2 = void 0;
const environments = __importStar(require("../../../../../../environments"));
const core = __importStar(require("../../../../../../core"));
const TwelvelabsApi = __importStar(require("../../../../../index"));
const serializers = __importStar(require("../../../../../../serialization/index"));
const url_join_1 = __importDefault(require("url-join"));
const errors = __importStar(require("../../../../../../errors/index"));
const Client_1 = require("../resources/tasks/client/Client");
class V2 {
    constructor(_options = {}) {
        this._options = _options;
    }
    get tasks() {
        var _a;
        return ((_a = this._tasks) !== null && _a !== void 0 ? _a : (this._tasks = new Client_1.Tasks(this._options)));
    }
    /**
     * This endpoint synchronously creates embeddings for multimodal content and returns the results immediately in the response.
     *
     * <Note title="Note">
     *   This method only supports Marengo version 3.0 or newer.
     * </Note>
     *
     * **When to use this endpoint**:
     * - Create embeddings for text, images, audio, or video content
     * - Get immediate results without waiting for background processing
     * - Process audio or video content up to 10 minutes in duration
     *
     * **Do not use this endpoint for**:
     * - Audio or video content longer than 10 minutes. Use the [`POST`](/v1.3/api-reference/create-embeddings-v2/create-async-embedding-task) method of the `/embed-v2/tasks` endpoint instead.
     *
     * <Accordion title="Input requirements">
     *   **Text**:
     *   - Maximum length: 500 tokens
     *
     *   **Images**:
     *   - Formats: JPEG, PNG
     *   - Minimum size: 128x128 pixels
     *   - Maximum file size: 5 MB
     *
     *   **Audio and video**:
     *   - Maximum duration: 10 minutes
     *   - Maximum file size for base64 encoded strings: 36 MB
     *   - Audio formats: WAV (uncompressed), MP3 (lossy), FLAC (lossless)
     *   - Video formats: [FFmpeg supported formats](https://ffmpeg.org/ffmpeg-formats.html)
     *   - Video resolution: 360x360 to 3840x2160 pixels
     *   - Aspect ratio: Between 1:1 and 1:2.4, or between 2.4:1 and 1:1
     * </Accordion>
     *
     * @param {TwelvelabsApi.embed.CreateEmbeddingsRequest} request
     * @param {V2.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link TwelvelabsApi.BadRequestError}
     * @throws {@link TwelvelabsApi.TooManyRequestsError}
     * @throws {@link TwelvelabsApi.InternalServerError}
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "text",
     *         modelName: "marengo3.0",
     *         text: {
     *             inputText: "man walking a dog"
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "image",
     *         modelName: "marengo3.0",
     *         image: {
     *             mediaSource: {
     *                 url: "https://user-bucket.com/folder/dog.jpg"
     *             }
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "text_image",
     *         modelName: "marengo3.0",
     *         textImage: {
     *             mediaSource: {
     *                 url: "https://user-bucket.com/folder/dog.jpg"
     *             },
     *             inputText: "man walking a dog"
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "image",
     *         modelName: "marengo3.0",
     *         image: {
     *             mediaSource: {
     *                 assetId: "1234567890"
     *             }
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "audio",
     *         modelName: "marengo3.0",
     *         audio: {
     *             mediaSource: {
     *                 url: "https://user-bucket.com/audio/a.wav"
     *             },
     *             startSec: 0,
     *             endSec: 6,
     *             segmentation: {
     *                 strategy: "fixed",
     *                 fixed: {
     *                     durationSec: 6
     *                 }
     *             },
     *             embeddingOption: ["audio", "transcription"],
     *             embeddingScope: ["clip", "asset"]
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "video",
     *         modelName: "marengo3.0",
     *         video: {
     *             mediaSource: {
     *                 url: "https://user-bucket.com/video/clip.mp4"
     *             },
     *             startSec: 0,
     *             endSec: 12,
     *             segmentation: {
     *                 strategy: "dynamic",
     *                 dynamic: {
     *                     minDurationSec: 4
     *                 }
     *             },
     *             embeddingOption: ["visual", "audio", "transcription"],
     *             embeddingScope: ["clip", "asset"]
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "video",
     *         modelName: "marengo3.0",
     *         video: {
     *             mediaSource: {
     *                 url: "https://user-bucket.com/video/simple.mp4"
     *             }
     *         }
     *     })
     *
     * @example
     *     await client.embed.v2.create({
     *         inputType: "audio",
     *         modelName: "marengo3.0",
     *         audio: {
     *             mediaSource: {
     *                 url: "https://user-bucket.com/audio/speech.wav"
     *             },
     *             embeddingOption: ["transcription"],
     *             embeddingScope: ["asset"]
     *         }
     *     })
     */
    create(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__create(request, requestOptions));
    }
    __create(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const _response = yield core.fetcher({
                url: (0, url_join_1.default)((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.TwelvelabsApiEnvironment.Default, "embed-v2"),
                method: "POST",
                headers: Object.assign(Object.assign({ "X-Fern-Language": "JavaScript", "X-Fern-SDK-Name": "twelvelabs-js", "X-Fern-SDK-Version": "1.1.0", "User-Agent": "twelvelabs-js/1.1.0", "X-Fern-Runtime": core.RUNTIME.type, "X-Fern-Runtime-Version": core.RUNTIME.version }, (yield this._getCustomAuthorizationHeaders())), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                contentType: "application/json",
                requestType: "json",
                body: serializers.embed.CreateEmbeddingsRequest.jsonOrThrow(request, { unrecognizedObjectKeys: "strip" }),
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return {
                    data: serializers.EmbeddingSuccessResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new TwelvelabsApi.BadRequestError(_response.error.body, _response.rawResponse);
                    case 429:
                        throw new TwelvelabsApi.TooManyRequestsError(_response.error.body, _response.rawResponse);
                    case 500:
                        throw new TwelvelabsApi.InternalServerError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TwelvelabsApiError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TwelvelabsApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TwelvelabsApiTimeoutError("Timeout exceeded when calling POST /embed-v2.");
                case "unknown":
                    throw new errors.TwelvelabsApiError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    _getCustomAuthorizationHeaders() {
        return __awaiter(this, void 0, void 0, function* () {
            const apiKeyValue = yield core.Supplier.get(this._options.apiKey);
            return { "x-api-key": apiKeyValue };
        });
    }
}
exports.V2 = V2;
